Source/Methods/Envelope Tracking/EnvelopeTracking.cs

```diff
@@ -0,0 +1,309 @@
+            WritableBuffer<ComplexSingle> envWfmWriteBuffer = envelopeWaveform.Data.GetWritableBuffer();
+
+            double[] iqMagnitudes = referenceWaveform.Data.GetMagnitudeDataArray(false);
+            double detroughRatio = detroughConfig.MinimumVoltage / detroughConfig.MaximumVoltage;
+
+            // Waveforms are assumed to be normalized in range [0, 1], so no normalization will happen here
+            switch (detroughConfig.Type)
+            {
+                case DetroughType.Exponential:
+                    // Formula: e = IQmag + d*exp(-IQmag/d)
+                    double expScale = 1 + detroughRatio * Math.Exp(-1.0 / detroughRatio);
+                    for (int i = 0; i < envWfmWriteBuffer.Size; i++)
+                    {
+                        double sampleValue = iqMagnitudes[i] + detroughRatio * Math.Exp(-iqMagnitudes[i] / detroughRatio);
+                        envWfmWriteBuffer[i] = ComplexSingle.FromSingle((float)(detroughConfig.MaximumVoltage * sampleValue / expScale)); // Scale detrough to Vmax. Divide by detroughed waveform's max value to normalize. IQMagnitude's max value is 1
+                    }
+                    break;
+                case DetroughType.Cosine:
+                    // Formula: e = 1 - (1-d)*cos(IQmag*pi/2)
+                    double cosScale = 1 - (1 - detroughRatio) * Math.Cos(Math.PI / 2);
+                    for (int i = 0; i < envWfmWriteBuffer.Size; i++)
+                    {
+                        double sampleValue = 1 - (1 - detroughRatio) * Math.Cos(iqMagnitudes[i] * cosScale);
+                        envWfmWriteBuffer[i] = ComplexSingle.FromSingle((float)(detroughConfig.MaximumVoltage * sampleValue / cosScale)); // Scale detrough to Vmax. Divide by detroughed waveform's max value to normalize. IQMagnitude's max value is 1: 1-(1-d) *cos(1*pi/2)
+                    }
+                    break;
+                case DetroughType.Power:
+                    // Formula: e = (1-d) + IQmag^(exponent)*(1-d)
+                    double powScale = 2 - 2 * detroughRatio;
+                    for (int i = 0; i < envWfmWriteBuffer.Size; i++)
+                    {
+                        double sampleValue = (1 - detroughRatio) + (Math.Pow(iqMagnitudes[i], detroughConfig.Exponent) * (1 - detroughRatio));
+                        envWfmWriteBuffer[i] = ComplexSingle.FromSingle((float)(detroughConfig.MaximumVoltage * sampleValue / powScale)); // Scale detrough to Vmax. Divide by detroughed waveform's max value. IQMagnitude's max value is 1: (1-d) + (1^exponent)*(1-d)
+                    }
+                    break;
+                default:
+                    throw new ArgumentException("Detrough type not supported.\n");
+            }
+
+            return envelopeWaveform;
+        }
+
+        public static Waveform CreateEnvelopeWaveform(Waveform referenceWaveform, LookUpTable lookUpTable, double dutAverageInputPower)

```

Comment:
If we renamed `LookupTable` to `LookupTableConfiguration` I think it would be reasonable to add  `dutAverageInputPower` as an element of that struct rather than a standalone input.

Reply:
05bde1e38bef31e3a9710d772050a815b59e148e